fitlins/interfaces/nistats.py

```diff
@@ -134,14 +134,50 @@ def _run_interface(self, runtime):
         # Save model level images
         model_maps = []
         model_metadata = []
-        for output in ['r_square']:
-            model_metadata.append(
-                {'stat': output,
-                 **out_ents}
-                )
-            fname = fname_fmt('model', output)
-            getattr(flm, output)[0].to_filename(fname)
-            model_maps.append(fname)
+
+        # R-square
+        model_metadata.append(
+            {'stat': 'r_square',
+             **out_ents}
+            )
+        fname = fname_fmt('model', 'r_square')
+        flm.r_square[0].to_filename(fname)
+        model_maps.append(fname)
+
+        # Manually compute log-likelihood
+
+        def _get_voxelwise_logL(flm):
```

Comment:
I don't know if it would be easier, but you could also patch the `FirstLevelModel` class with a more correct `_get_voxelwise_model_attribute` function definition, and use [`FirstLevelModelPatch` here](https://github.com/poldracklab/fitlins/pull/230/files#diff-efda2d1eedc98c2ee14427f38cf64710L126).
The main benefit being if this change is pulled into nilearn/nistats, then you can just delete the patched class and use the "normal" one again and not potentially change how you are solving the problem (by defining a custom function `_get_voxelwise_logL`).

Overthinking this since your solution works and is also good (and ultimately fewer lines of code).


```
from nistats.regression import RegressionResults
from nistats.model import LikelihoodModelResults

class FirstLevelModelPatch(FirstLevelModel)
    def _get_voxelwise_model_attribute(self, attribute,
                                       result_as_time_series):
        """Transform RegressionResults instances within a dictionary
        (whose keys represent the autoregressive coefficient under the 'ar1'
        noise model or only 0.0 under 'ols' noise_model and values are the
        RegressionResults instances) into input nifti space.
        Parameters
        ----------
        attribute : str
            an attribute of a RegressionResults instance.
            possible values include: resid, norm_resid, predicted,
            SSE, r_square, MSE.
        result_as_time_series : bool
            whether the RegressionResult attribute has a value
            per timepoint of the input nifti image.
        Returns
        -------
        output : list
            a list of Nifti1Image(s)
        """
        # check if valid attribute is being accessed.
         # CHANGING THIS LINE TO BE BETTER
        all_attributes = set(dict(vars(LikelihoodModelResults)).keys()).union(dict(vars(RegressionResults)).keys())
        possible_attributes = [prop
                               for prop in all_attributes
                               if '__' not in prop
                               ]
        if attribute not in possible_attributes:
            msg = ("attribute must be one of: "
                   "{attr}".format(attr=possible_attributes)
                   )
            raise ValueError(msg)

        if self.minimize_memory:
            raise ValueError(
                'To access voxelwise attributes like '
                'R-squared, residuals, and predictions, '
                'the `FirstLevelModel`-object needs to store '
                'there attributes. '
                'To do so, set `minimize_memory` to `False` '
                'when initializing the `FirstLevelModel`-object.')

        if self.labels_ is None or self.results_ is None:
            raise ValueError('The model has not been fit yet')

        output = []

        for design_matrix, labels, results in zip(self.design_matrices_,
                                                  self.labels_,
                                                  self.results_
                                                  ):
            if result_as_time_series:
                voxelwise_attribute = np.zeros((design_matrix.shape[0],
                                                len(labels))
                                               )
            else:
                voxelwise_attribute = np.zeros((1, len(labels)))

            for label_ in results:
                label_mask = labels == label_
                voxelwise_attribute[:, label_mask] = getattr(results[label_],
                                                             attribute)

            output.append(self.masker_.inverse_transform(voxelwise_attribute))

            return output
```

Reply:
I think subclassing things from nistats is probably going to be more effort than it's worth at this point, but worth thinking about making these contributions upstream to nilearn post-0.7.0.

@adelavega Can you define this function at the file-level, rather than nested inside this scope?

Reply:
Yeah, I can do that. I think that's easiest for now.

Reply:
Sounds good to me, I can reference an issue in nilearn.